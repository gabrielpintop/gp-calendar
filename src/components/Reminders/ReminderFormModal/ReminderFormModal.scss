.label {
    font-weight: bold;
}

.form-control {
    display: block;
    background-color: transparent;
    border: 1px solid #4D4352;
    border-radius: 0.25rem;
    font-size: 16px;
    padding: 5px 5px;
    outline: none !important;
    width: 270px;
    margin-top: 6px;
    margin-bottom: 10px;
}

#closeHeader {
    text-align: right;
    cursor: pointer;
}

.weather-information {
    display: flex;
    font-size: 0.9rem;
    font-weight: bold;
    color: #4E81A5;
    align-items: center;

    .loading-weather {
        margin-bottom: 1rem;
        font-size: 0.8rem;
    }

    img {
        width: 40px;
    }
}