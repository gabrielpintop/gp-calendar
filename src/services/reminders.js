export const addReminder = (year, month, day, { id, text, time, city, color }) => {
    if (year === null || month === null || day === null) {
        return 'The dates of the reminder are not valid';
    } else if (!text) {
        return 'The text of the reminder is required';
    } else if (text.length > 30) {
        return 'Reminder length is not valid';
    } else if (!time) {
        return 'The time of the reminder is required';
    } else if (!city) {
        return 'The city of the reminder is required';
    } else if (!color) {
        return 'The color of the reminder is required';
    }

    let reminders = {};
    try {
        const parsedReminders = JSON.parse(localStorage.getItem('GP_CALENDAR_STUFF'));
        if (parsedReminders) {
            reminders = parsedReminders;
        }
    } catch (error) {
        console.log(error);
    }

    const reminder = {
        id: id || new Date().getTime(),
        text,
        time,
        city,
        color
    };

    if (!reminders[year]) {
        reminders[year] = {
            [month]: {
                [day]: [reminder]
            }
        };
    } else if (!reminders[year][month]) {
        reminders[year][month] = {
            [day]: [reminder]
        }
    } else if (!reminders[year][month][day]) {
        reminders[year][month][day] = [reminder];
    } else {
        reminders[year][month][day].push(reminder);
        reminders[year][month][day].sort((a, b) => (a.time > b.time) ? 1 : ((b.time > a.time) ? -1 : 0));
    }

    localStorage.setItem('GP_CALENDAR_STUFF', JSON.stringify(reminders));
    return true;
};

export const getRemindersByYearAndMonth = (year, month) => {
    const reminders = JSON.parse(localStorage.getItem('GP_CALENDAR_STUFF'));
    if (reminders && reminders[year] && reminders[year][month]) {
        return reminders[year][month];
    } else {
        return {};
    }
};

export const updateReminder = (year, month, day, reminder) => {
    if (deleteReminder(year, month, day, reminder.id)) {
        return addReminder(reminder.year, reminder.month, reminder.day, reminder);
    } else {
        return false;
    }
};

export const deleteReminder = (year, month, day, id) => {
    try {
        const parsedReminders = JSON.parse(localStorage.getItem('GP_CALENDAR_STUFF'));
        if (parsedReminders) {
            const reminders = parsedReminders[year][month][day];
            const index = reminders.findIndex(reminder => reminder.id === id);
            if (index === -1) {
                return false;
            } else {

                reminders.splice(index, 1);
                parsedReminders[year][month][day] = reminders;
                localStorage.setItem('GP_CALENDAR_STUFF', JSON.stringify(parsedReminders));
                return true;
            }
        } else {
            return false;
        }
    } catch (error) {
        console.log(error);
        return false;
    }
};

export const deleteReminders = (year, month, day) => {
    try {
        const parsedReminders = JSON.parse(localStorage.getItem('GP_CALENDAR_STUFF'));
        if (parsedReminders) {
            parsedReminders[year][month][day] = [];
            localStorage.setItem('GP_CALENDAR_STUFF', JSON.stringify(parsedReminders));
            return true;
        } else {
            return false;
        }
    } catch (error) {
        console.log(error);
        return false;
    }
};

export const getRemindersByYearMonthAndDay = (year, month, day) => {
    const reminders = getRemindersByYearAndMonth(year, month);

    if (reminders) {
        return reminders[day] || [];
    }
};