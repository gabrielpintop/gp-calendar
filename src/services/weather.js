const axios = require('axios').default;
const api = 'https://api.openweathermap.org/data/2.5/weather?appid=d2ec863eaa2321925fed4d433208447b&q=';

const getWeatherBasedOnCity = (city) => {
    return new Promise((resolve, reject) => {
        axios.get(`${api}${city}`).then(({ data }) => {
            if (data.weather && data.weather[0]) {
                resolve(data.weather[0]);
            } else {
                reject(false);
            }
        }).catch(err => {
            console.log(err);
            reject(false);
        });
    });
};

module.exports = {
    getWeatherBasedOnCity
}
