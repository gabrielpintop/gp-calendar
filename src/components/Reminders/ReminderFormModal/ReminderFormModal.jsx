import React, { useState, useEffect } from 'react';
import Modal from 'react-modal';
import './ReminderFormModal.scss';
import { connect } from 'react-redux';
import { toggleModal, setRemindersAndDate } from '../../../actions';
import { getWeatherBasedOnCity } from '../../../services/weather';
import { addReminder, updateReminder, deleteReminder } from '../../../services/reminders';

const ReminderForm = (props) => {

    const { reminder, showModal, toggleModal, year, month, day, setRemindersAndDate } = props;

    const [loading, setLoading] = useState(false);

    const [weatherData, setWeatherData] = useState(null);

    const customStyles = {
        content: {
            top: '50%',
            left: '50%',
            right: 'auto',
            bottom: 'auto',
            marginRight: '-50%',
            transform: 'translate(-50%, -50%)',
            borderColor: '#a42ef3'
        }
    };

    const [form, setValues] = useState({
        text: reminder.text || '',
        time: reminder.time || new Date().toLocaleTimeString().slice(0, 5),
        date: new Date(year, month, day).toISOString().substr(0, 10),
        city: reminder.city || '',
        color: reminder.color || '#009D87'
    });

    const handleInput = event => {
        setValues({
            ...form,
            [event.target.name]: event.target.value
        });
    };

    const getCityWeather = () => {
        setLoading(true);
        getWeatherBasedOnCity(reminder.city).then(data => {
            setWeatherData(data);
            setLoading(false);
        }).catch(err => {
            setLoading(false);
        });
    };

    const handleSubmit = event => {
        event.preventDefault();
        const dateElements = form.date.split('-');
        if (reminder.id) {
            updateExistingReminder(dateElements);
        } else {
            addNewReminder(dateElements);
        }
    };

    const addNewReminder = (dateElements) => {
        const result = addReminder(Number(dateElements[0]), Number(dateElements[1]) - 1, Number(dateElements[2]), form);
        if (result === true) {
            setRemindersAndDate({ year, month, day, showModal: false, reminder: {} });
        } else {
            window.alert(result);
        }
    };

    const updateExistingReminder = (dateElements) => {
        const result = updateReminder(year, month, day, { ...form, id: reminder.id, year: Number(dateElements[0]), month: Number(dateElements[1]) - 1, day: Number(dateElements[2]) });
        if (result) {
            setRemindersAndDate({ year, month, day, showModal: false, reminder: {} });
        } else {
            window.alert('There was an error while updating the reminder');
        }
    };

    const handleDeleteReminder = () => {
        if (deleteReminder(year, month, day, reminder.id)) {
            setRemindersAndDate({ year, month, day, showModal: false, reminder: {} });
        } else {
            window.alert('There was an error while deleting the reminder')
        }
    };

    useEffect(() => {
        if (reminder.id && reminder.city) {
            getCityWeather();
        }
    }, []);

    return (
        <Modal isOpen={showModal} style={customStyles}>
            <div id="closeHeader"><i className="fas fa-times" onClick={() => toggleModal({ showModal: false, reminder: {} })}></i></div>
            {reminder.id && <div className="weather-information">{loading ? <span className="loading-weather">Loading {reminder.city} weather...</span> : weatherData && <><img src={`http://openweathermap.org/img/wn/${weatherData.icon}@2x.png`} alt={weatherData.main}></img><span>{weatherData.main} in {reminder.city}</span></>}</div>}
            <form onSubmit={handleSubmit}>
                <label className="label" htmlFor="text">Text</label>
                <input
                    id="text"
                    name="text"
                    className="form-control"
                    type="text"
                    value={form.text}
                    onChange={handleInput}
                    minLength="1"
                    maxLength="30"
                    required
                />
                <label className="label" htmlFor="date">Date</label>
                <input
                    id="date"
                    name="date"
                    className="form-control"
                    type="date"
                    value={form.date}
                    onChange={handleInput}
                    required
                />
                <label className="label" htmlFor="text">Time</label>
                <input
                    id="time"
                    name="time"
                    className="form-control"
                    type="time"
                    value={form.time}
                    onChange={handleInput}
                    required
                />

                <label className="label" htmlFor="city">City</label>
                <input
                    id="city"
                    name="city"
                    className="form-control"
                    type="text"
                    value={form.city}
                    onChange={handleInput}
                    maxLength="50"
                    required
                />
                <label className="label" htmlFor="color">Color</label>
                <input
                    id="color"
                    name="color"
                    className="form-control"
                    type="color"
                    value={form.color}
                    onChange={handleInput}
                    required
                />
                <small>*Do not choose light colors for better usability</small>
                <br />
                <button type="submit" className="button-add margin-top-1 "><i className={`fas ${reminder.id ? 'fa-save' : 'fa-plus'}`} ></i>&nbsp;{reminder.id ? 'Save' : 'Create'}&nbsp;</button>
                {reminder.id && <button className="button-delete margin-top-1 float-right" type="button" onClick={handleDeleteReminder}><i className="fas fa-trash" ></i>&nbsp;Delete&nbsp;</button>}
            </form>
        </Modal>
    );
};

const mapStateToProps = ({ showModal, reminder, year, month, day }) => {
    return {
        showModal,
        reminder,
        year,
        month,
        day
    };
};

const mapDispatchToProps = {
    toggleModal,
    setRemindersAndDate
};

export default connect(mapStateToProps, mapDispatchToProps)(ReminderForm);