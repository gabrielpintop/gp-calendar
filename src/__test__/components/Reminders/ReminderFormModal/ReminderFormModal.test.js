import React from 'react';
import { mount } from 'enzyme';
import ReminderFormModal from '../../../../components/Reminders/ReminderFormModal/ReminderFormModal';
import Modal from 'react-modal';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import * as reminders from '../../../../services/reminders';
import EnzymeToJson from 'enzyme-to-json';

describe('ReminderFormModal component', () => {

    let store;
    let reminderFormModal;

    beforeAll(() => {
        const mockStore = configureStore();
        Modal.setAppElement('body');
        store = mockStore({
            showModal: true,
            reminder: {
                time: '15:00'
            },
            year: 2020,
            month: 1,
            day: 21,
            reminders: []
        });
    });

    beforeEach(() => {
        reminderFormModal = mount(<Provider store={store}><ReminderFormModal /></Provider>);
    });

    test('should match the snapshot', () => {
        expect(EnzymeToJson(reminderFormModal)).toMatchSnapshot();
    });

    test('should render', () => {
        expect(reminderFormModal.length).toEqual(1);
    });

    test('should have text input', () => {
        expect(reminderFormModal.find('#text').props().type).toEqual('text');
    });

    test('should have date input', () => {
        expect(reminderFormModal.find('#date').props().type).toEqual('date');
    });

    test('should have time input', () => {
        expect(reminderFormModal.find('#time').props().type).toEqual('time');
    });

    test('should have city input', () => {
        expect(reminderFormModal.find('#city').props().type).toEqual('text');
    });

    test('should have color input', () => {
        expect(reminderFormModal.find('#color').props().type).toEqual('color');
    });

    test('should have create button', () => {
        expect(reminderFormModal.find('button').length).toEqual(1);
    });

    test('should not call the method to submit calendar', () => {
        reminders.addReminder = jest.fn();
        reminderFormModal.find('form').simulate('submit');
        expect(reminders.addReminder).toHaveBeenCalledTimes(1);
    });

    test('should call the method to submit calendar', () => {
        reminders.addReminder = jest.fn();
        reminderFormModal.find('#text').instance().value = 'Add a reminder test';
        reminderFormModal.find('#text').simulate('change');
        reminderFormModal.find('#city').instance().value = 'Bogota';
        reminderFormModal.find('#city').simulate('change');
        reminderFormModal.find('form').simulate('submit');
        expect(reminders.addReminder).toHaveBeenCalledTimes(1);
    });
});