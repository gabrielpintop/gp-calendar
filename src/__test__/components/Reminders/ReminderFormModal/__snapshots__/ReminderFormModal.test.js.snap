// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReminderFormModal component should match the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ReminderForm)>
    <ReminderForm
      day={21}
      month={1}
      reminder={
        Object {
          "time": "15:00",
        }
      }
      setRemindersAndDate={[Function]}
      showModal={true}
      toggleModal={[Function]}
      year={2020}
    >
      <Modal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        closeTimeoutMS={0}
        isOpen={true}
        parentSelector={[Function]}
        portalClassName="ReactModalPortal"
        role="dialog"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={true}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
        style={
          Object {
            "content": Object {
              "borderColor": "#a42ef3",
              "bottom": "auto",
              "left": "50%",
              "marginRight": "-50%",
              "right": "auto",
              "top": "50%",
              "transform": "translate(-50%, -50%)",
            },
          }
        }
      >
        <Portal
          containerInfo={
            <div
              class="ReactModalPortal"
            >
              <div
                class="ReactModal__Overlay ReactModal__Overlay--after-open"
                style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
              >
                <div
                  class="ReactModal__Content ReactModal__Content--after-open"
                  role="dialog"
                  style="position: absolute; top: 50%; left: 50%; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px; margin-right: -50%; transform: translate(-50%, -50%); border-color: #a42ef3;"
                  tabindex="-1"
                >
                  <div
                    id="closeHeader"
                  >
                    <i
                      class="fas fa-times"
                    />
                  </div>
                  <form>
                    <label
                      class="label"
                      for="text"
                    >
                      Text
                    </label>
                    <input
                      class="form-control"
                      id="text"
                      maxlength="30"
                      minlength="1"
                      name="text"
                      required=""
                      type="text"
                      value=""
                    />
                    <label
                      class="label"
                      for="date"
                    >
                      Date
                    </label>
                    <input
                      class="form-control"
                      id="date"
                      name="date"
                      required=""
                      type="date"
                      value="2020-02-21"
                    />
                    <label
                      class="label"
                      for="text"
                    >
                      Time
                    </label>
                    <input
                      class="form-control"
                      id="time"
                      name="time"
                      required=""
                      type="time"
                      value="15:00"
                    />
                    <label
                      class="label"
                      for="city"
                    >
                      City
                    </label>
                    <input
                      class="form-control"
                      id="city"
                      maxlength="50"
                      name="city"
                      required=""
                      type="text"
                      value=""
                    />
                    <label
                      class="label"
                      for="color"
                    >
                      Color
                    </label>
                    <input
                      class="form-control"
                      id="color"
                      name="color"
                      required=""
                      type="color"
                      value="#009D87"
                    />
                    <small>
                      *Do not choose light colors for better usability
                    </small>
                    <br />
                    <button
                      class="button-add margin-top-1 "
                      type="submit"
                    >
                      <i
                        class="fas fa-plus"
                      />
                       
                      Create
                       
                    </button>
                  </form>
                </div>
              </div>
            </div>
          }
        >
          <ModalPortal
            ariaHideApp={true}
            bodyOpenClassName="ReactModal__Body--open"
            closeTimeoutMS={0}
            defaultStyles={
              Object {
                "content": Object {
                  "WebkitOverflowScrolling": "touch",
                  "background": "#fff",
                  "border": "1px solid #ccc",
                  "borderRadius": "4px",
                  "bottom": "40px",
                  "left": "40px",
                  "outline": "none",
                  "overflow": "auto",
                  "padding": "20px",
                  "position": "absolute",
                  "right": "40px",
                  "top": "40px",
                },
                "overlay": Object {
                  "backgroundColor": "rgba(255, 255, 255, 0.75)",
                  "bottom": 0,
                  "left": 0,
                  "position": "fixed",
                  "right": 0,
                  "top": 0,
                },
              }
            }
            isOpen={true}
            parentSelector={[Function]}
            portalClassName="ReactModalPortal"
            role="dialog"
            shouldCloseOnEsc={true}
            shouldCloseOnOverlayClick={true}
            shouldFocusAfterRender={true}
            shouldReturnFocusAfterClose={true}
            style={
              Object {
                "content": Object {
                  "borderColor": "#a42ef3",
                  "bottom": "auto",
                  "left": "50%",
                  "marginRight": "-50%",
                  "right": "auto",
                  "top": "50%",
                  "transform": "translate(-50%, -50%)",
                },
              }
            }
          >
            <div
              className="ReactModal__Overlay ReactModal__Overlay--after-open"
              onClick={[Function]}
              onMouseDown={[Function]}
              style={
                Object {
                  "backgroundColor": "rgba(255, 255, 255, 0.75)",
                  "bottom": 0,
                  "left": 0,
                  "position": "fixed",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <div
                className="ReactModal__Content ReactModal__Content--after-open"
                onClick={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                role="dialog"
                style={
                  Object {
                    "WebkitOverflowScrolling": "touch",
                    "background": "#fff",
                    "border": "1px solid #ccc",
                    "borderColor": "#a42ef3",
                    "borderRadius": "4px",
                    "bottom": "auto",
                    "left": "50%",
                    "marginRight": "-50%",
                    "outline": "none",
                    "overflow": "auto",
                    "padding": "20px",
                    "position": "absolute",
                    "right": "auto",
                    "top": "50%",
                    "transform": "translate(-50%, -50%)",
                  }
                }
                tabIndex="-1"
              >
                <div
                  id="closeHeader"
                >
                  <i
                    className="fas fa-times"
                    onClick={[Function]}
                  />
                </div>
                <form
                  onSubmit={[Function]}
                >
                  <label
                    className="label"
                    htmlFor="text"
                  >
                    Text
                  </label>
                  <input
                    className="form-control"
                    id="text"
                    maxLength="30"
                    minLength="1"
                    name="text"
                    onChange={[Function]}
                    required={true}
                    type="text"
                    value=""
                  />
                  <label
                    className="label"
                    htmlFor="date"
                  >
                    Date
                  </label>
                  <input
                    className="form-control"
                    id="date"
                    name="date"
                    onChange={[Function]}
                    required={true}
                    type="date"
                    value="2020-02-21"
                  />
                  <label
                    className="label"
                    htmlFor="text"
                  >
                    Time
                  </label>
                  <input
                    className="form-control"
                    id="time"
                    name="time"
                    onChange={[Function]}
                    required={true}
                    type="time"
                    value="15:00"
                  />
                  <label
                    className="label"
                    htmlFor="city"
                  >
                    City
                  </label>
                  <input
                    className="form-control"
                    id="city"
                    maxLength="50"
                    name="city"
                    onChange={[Function]}
                    required={true}
                    type="text"
                    value=""
                  />
                  <label
                    className="label"
                    htmlFor="color"
                  >
                    Color
                  </label>
                  <input
                    className="form-control"
                    id="color"
                    name="color"
                    onChange={[Function]}
                    required={true}
                    type="color"
                    value="#009D87"
                  />
                  <small>
                    *Do not choose light colors for better usability
                  </small>
                  <br />
                  <button
                    className="button-add margin-top-1 "
                    type="submit"
                  >
                    <i
                      className="fas fa-plus"
                    />
                     
                    Create
                     
                  </button>
                </form>
              </div>
            </div>
          </ModalPortal>
        </Portal>
      </Modal>
    </ReminderForm>
  </Connect(ReminderForm)>
</Provider>
`;
